drivers/vl53l1x/vl53l1x.c

```diff
@@ -0,0 +1,1316 @@
+        dev->int_init = true;
+        gpio_init_int(dev->params.pin_int, GPIO_IN_PU, GPIO_FALLING,
+                      isr, isr_arg);
+    }
+
+#if IS_USED(MODULE_VL53L1X_ST_API)
+    ASSERT_PARAM(cfg != NULL);
+
+    EXEC_RET(VL53L1_StopMeasurement(&dev->dev));
+
+    VL53L1_DetectionConfig_t dcfg;
+    dcfg.DetectionMode = cfg->mode;
+    dcfg.Distance.CrossMode = cfg->dist_mode;
+    dcfg.Distance.High = cfg->dist_high;
+    dcfg.Distance.Low = cfg->dist_low;
+    dcfg.Rate.CrossMode = cfg->rate_mode;
+    dcfg.Rate.High = cfg->rate_high;
+    dcfg.Rate.Low = cfg->rate_low;
+    dcfg.IntrNoTarget = 0;
+
+    EXEC_RET(VL53L1_SetThresholdConfig(&dev->dev, &dcfg));
+    EXEC_RET(VL53L1_StartMeasurement(&dev->dev));
+#else
+    (void)cfg;
+    DEBUG_DEV("configuration is not used", dev);
+#endif
+
+    return VL53L1X_OK;
+}
+
+#if !IS_USED(MODULE_VL53L1X_BASIC)
+
+int vl53l1x_power_down(const vl53l1x_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    if (dev->params.pin_shutdown == GPIO_UNDEF) {
+        DEBUG_DEV("Pin connected to sensor's XSHUT pin not defined", dev);
+        return VL53L1X_ERROR_NO_PIN;
+    }
+
+    gpio_clear(dev->params.pin_shutdown);
+    return VL53L1X_OK;
+}
+
+int vl53l1x_power_up(vl53l1x_t *dev)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("", dev);
+
+    if (dev->params.pin_shutdown == GPIO_UNDEF) {
+        DEBUG_DEV("Pin connected to sensor's XSHUT pin not defined", dev);
+        return VL53L1X_ERROR_NO_PIN;
+    }
+
+    gpio_set(dev->params.pin_shutdown);
+
+    /* init the sensor and start measurement */
+    _init(dev);
+
+    return VL53L1X_OK;
+}
+
+int vl53l1x_get_timing_budget(vl53l1x_t *dev, uint32_t *budget)
+{
+    ASSERT_PARAM(dev != NULL);
+    ASSERT_PARAM(budget != NULL);
+    DEBUG_DEV("budget=%p", dev, budget);
+
+#if IS_USED(MODULE_VL53L1X_ST_API)
+    return VL53L1_GetMeasurementTimingBudgetMicroSeconds(&dev->dev,
+                                                         budget);
+#else
+    /* Copyrighted POLOLU / ST API code begin */
+
+    /* Update Macro Period for Range A VCSEL Period */
+    uint8_t  byte;
+    uint32_t period_us;
+
+    EXEC_RET(_read_byte(dev, VL53L1X_RANGE_CONFIG__VCSEL_PERIOD_A, &byte));
+    period_us = _calc_macro_period(dev, byte);
+
+    /* Get Range Timing A timeout */
+    uint16_t word;
+    uint32_t timeout_us;
+
+    EXEC_RET(_read_word(dev, VL53L1X_RANGE_CONFIG__TIMEOUT_MACROP_A, &word));
+    timeout_us = _timeout_clk_to_us(_dec_timeout(word), period_us);
+
+    *budget = 2 * timeout_us + _timing_guard;
+    /* Copyrighted POLOLU / ST API code end */
+
+    return VL53L1X_OK;
+
+#endif
+}
+
+int vl53l1x_set_timing_budget(vl53l1x_t *dev, uint32_t budget_us)
+{
+    ASSERT_PARAM(dev != NULL);
+    DEBUG_DEV("budget_us=%u", dev, budget_us);
+
+#if IS_USED(MODULE_VL53L1X_ST_API)
+
+    EXEC_RET(VL53L1_SetMeasurementTimingBudgetMicroSeconds(&dev->dev,
+                                                           budget_us));
+#else
```

Comment:
Is there a chance that this would work as C conditionals? That way we can be sure that both ST-API and pololu variant can be compiled without having to provide two test applications. (More instances above and below.)

Reply:
Hm, there are a lot of defines and symbol mappings dependent on whether the ST API is used or not that I can't imaging that compilation works if we use C conditionals. I can try it, but I'm not optimistic.

Reply:
I think there is no chance in this case. The problem is that the ST API is used as a package, so of course the header files with a lot of defines and declarations required during compilation are not available if the package is not used by the driver:
```python
drivers/vl53l1x/vl53l1x.c:525:14: error: implicit declaration of function 'VL53L1_SetMeasurementTimingBudgetMicroSeconds' [-Werror=implicit-function-declaration]
     EXEC_RET(VL53L1_SetMeasurementTimingBudgetMicroSeconds(&dev->dev,
              ^
```
To solve this problem we would have to pull the package in any case and insert the headers always, even if they are not used by the driver. IMHO, that would be the wrong approach just to use C Conditionals.

Even if we would do it in that way, there would raise another compilation problem because the driver defines some mappings to be able to use same code independent on whether the ST API is used or not. These mappings would result in a conflict if the ST API is used.

I fully agree that the CI coverage is an argument to use C conditionals. But in this case I would prefer to keep the preprocessor conditionals and to provide a second test application. Would it be ok for you?